src/webapp/components/form/FormState.ts

```diff
@@ -81,50 +85,215 @@ export type FormFieldState =
+            } else {
+                return [...acc, ...section.fields];
+            }
+        },
+        []
+    );
+}
+
+export function getStringFieldValue(id: string, allFields: FormFieldState[]): string {
+    return (
+        getFieldValueById<FormTextFieldState | FormOptionsFieldState | FormAvatarFieldState>(
+            id,
+            allFields
+        ) || ""
+    );
+}
+
+export function getDateFieldValue(id: string, allFields: FormFieldState[]): Date | null {
+    return getFieldValueById<FormDateFieldState>(id, allFields) || null;
+}
+
+export function getBooleanFieldValue(id: string, allFields: FormFieldState[]): boolean {
+    return !!getFieldValueById<FormBooleanFieldState>(id, allFields);
+}
+
+export function getMultipleOptionsFieldValue(id: string, allFields: FormFieldState[]): string[] {
+    return getFieldValueById<FormMultipleOptionsFieldState>(id, allFields) || [];
+}
+
+export function getFieldValueById<F extends FormFieldState>(
+    id: string,
+    fields: FormFieldState[]
+): F["value"] | undefined {
+    const field = fields.find(field => field.id === id);
+    if (field) {
+        return getFieldValue<F>(field);
+    }
+}
+
+export function getFieldIdFromIdsDictionary<T extends Record<string, string>>(
+    key: keyof T,
+    fieldIdsDictionary: T
+): string {
+    return fieldIdsDictionary[key] as string;
+}
+
+function getEmptyValueForField<F extends FormFieldState>(field: FormFieldState): F["value"] {
+    switch (field.type) {
+        case "text":
+            return "";
+        case "boolean":
+            return false;
+        case "select":
+            return field.multiple ? [] : "";
+        case "radio":
+            return "";
+        case "date":
+            return null;
+        case "user":
+            return undefined;
+    }
+}
+
+function getFieldValue<F extends FormFieldState>(field: FormFieldState): F["value"] {
+    return field.value;
+}
+
+// UPDATES:
+
+export function updateFormStateAndApplySideEffects(
+    currentFormState: FormState,
+    updatedField: FormFieldState
+): FormState {
+    const updatedFormState = updateFormState(currentFormState, updatedField);
+
+    return applySideEffects(updatedFormState, updatedField);
+}
+
+function applySideEffects(updatedFormState: FormState, updatedField: FormFieldState): FormState {
+    if (updatedField.notApplicableFieldId) {
+        return {
+            ...applyEffectNotApplicableFieldUpdated(updatedFormState),
+        };
+    } else {
+        return updatedFormState;
+    }
+}
+
+function applyEffectNotApplicableFieldUpdated(formState: FormState): FormState {
+    return {
+        ...formState,
+        sections: applyEffectNotApplicableFieldUpdatedInSections(formState.sections),
+    };
+}
+
+function applyEffectNotApplicableFieldUpdatedInSections(
+    sectionsState: FormSectionState[]
+): FormSectionState[] {
+    return sectionsState.map(section => {
+        if (section?.subsections) {
+            return {
```

Comment:
again, this side effect needs to be applied only to the section which contains side-effect field, we can add that condition and avoid calling applyEffect for all sections.